jcl/src/java.base/share/classes/com/ibm/oti/util/Util.java

```diff
@@ -350,4 +383,58 @@ public static void printStackTraceElement(StackTraceElement e, Object elementSou
 	}
 }
 
+/*[IF JAVA_SPEC_VERSION >= 11]*/
+/**
+ * Helper method for printStackTraceElement to check if the classloader name
+ * should be included in the stack trace for the provided StackTraceElement.
+ *
+ * @param element The StackTraceElement to check
+ * @return true if the classloader name should be included, false otherwise
+ */
+private static boolean includeClassLoaderName(StackTraceElement element) {
+	return vmLangAccess.getIncludeClassLoaderName(element);
+}
+
+/**
+ * Helper method for printStackTraceElement to check if the module version
+ * should be included in the stack trace for the provided StackTraceElement.
+ *
+ * The module version should not be included for modules that are non-upgradeable.
+ *
+ * @param element The StackTraceElement to check
+ * @return true if the module version should be included, false otherwise
+ */
+private static boolean includeModuleVersion(StackTraceElement element) {
+	boolean includeModuleVersion = vmLangAccess.getIncludeModuleVersion(element);
+
+	if (includeModuleVersion) {
+		/* includeModuleVersion should only be set to true if the module is not a non-upgradeable module */
+		includeModuleVersion = !NonUpgradeableModules.moduleNames.contains(element.getModuleName());
+	}
+
+	return includeModuleVersion;
+}
+
+/**
+ * NonUpgradeableModules statically determines the set of non-upgradeable modules
+ * by checking the module names that are recorded in the hashes of the java.base
+ * module. If a module name is contained within this set, then the module is a
+ * non-upgradeable module. Non-upgradeable modules should not have their module
+ * version displayed in a stack trace if the calling class is Throwable or StackFrame.
+ */
+private static class NonUpgradeableModules {
+	static Set<String> moduleNames;
+
+	static {
+		Optional<ResolvedModule> javaBaseModule = ModuleLayer.boot().configuration().findModule("java.base"); //$NON-NLS-1$
+
+		if (javaBaseModule.isPresent()) {
```

Comment:
I'm not sure if java.base would ever not be present, but this code seems to anticipate that, in which case a NullPointerException would result on line 412. Might it be better to throw an Error here if java.base is not present?

Reply:
That sounds good. Something as simple as the following?

```
else {
	throw new Error("java.base module could not be found"); //$NON-NLS-1$
}
```

Reply:
or:
```
private static final class NonUpgradeableModules {
	static final Set<String> moduleNames;

	static {
		ResolvedModule resolvedJavaBaseModule = ModuleLayer.boot().configuration().findModule("java.base") //$NON-NLS-1$
				.orElseThrow(() -> new InternalError("java.base module could not be found")); //$NON-NLS-1$
		ModuleReferenceImpl javaBaseModuleRef = (ModuleReferenceImpl) resolvedJavaBaseModule.reference();

		moduleNames = javaBaseModuleRef.recordedHashes().names();
	}
}
```
Notice that both the class and the field can be `final`.